src/test/basics/Number_test.cpp

```diff
@@ -0,0 +1,468 @@
+            {Number{1'000'000'000'000'000, -15},
+             Number{1'000'000'000'000'000, -15},
+             Number{0}},
+            {Number{1'000'000'000'000'000, -15},
+             Number{1'000'000'000'000'001, -15},
+             Number{-1'000'000'000'000'000, -30}},
+            {Number{1'000'000'000'000'001, -15},
+             Number{1'000'000'000'000'000, -15},
+             Number{1'000'000'000'000'000, -30}}};
+        for (auto const& [x, y, z] : c)
+            BEAST_EXPECT(x - y == z);
+    }
+
+    void
+    test_mul()
+    {
+        testcase("test_mul");
+        using Case = std::tuple<Number, Number, Number>;
+        Case c[]{
+            {Number{7}, Number{8}, Number{56}},
+            {Number{1414213562373095, -15},
+             Number{1414213562373095, -15},
+             Number{2000000000000000, -15}},
+            {Number{-1414213562373095, -15},
+             Number{1414213562373095, -15},
+             Number{-2000000000000000, -15}},
+            {Number{-1414213562373095, -15},
+             Number{-1414213562373095, -15},
+             Number{2000000000000000, -15}},
+            {Number{3214285714285706, -15},
+             Number{3111111111111119, -15},
+             Number{1000000000000000, -14}},
+            {Number{1000000000000000, -32768},
+             Number{1000000000000000, -32768},
+             Number{0}}};
+        for (auto const& [x, y, z] : c)
+            BEAST_EXPECT(x * y == z);
+        bool caught = false;
+        try
+        {
+            Number{9'999'999'999'999'999, 32768} *
+                Number{5'000'000'000'000'000, 32767};
+        }
+        catch (std::overflow_error const&)
+        {
+            caught = true;
+        }
+        BEAST_EXPECT(caught);
+    }
+
+    void
+    test_div()
+    {
+        testcase("test_div");
+        using Case = std::tuple<Number, Number, Number>;
+        Case c[]{
+            {Number{1}, Number{2}, Number{5, -1}},
+            {Number{1}, Number{10}, Number{1, -1}},
+            {Number{1}, Number{-10}, Number{-1, -1}},
+            {Number{0}, Number{100}, Number{0}}};
+        for (auto const& [x, y, z] : c)
+            BEAST_EXPECT(x / y == z);
+        bool caught = false;
+        try
+        {
+            Number{1000000000000000, -15} / Number{0};
+        }
+        catch (std::overflow_error const&)
+        {
+            caught = true;
+        }
+        BEAST_EXPECT(caught);
+    }
+
+    void
+    test_root()
+    {
+        testcase("test_root");
+        using Case = std::tuple<Number, unsigned, Number>;
+        Case c[]{
+            {Number{2}, 2, Number{1414213562373095, -15}},
+            {Number{2'000'000}, 2, Number{1414213562373095, -12}},
+            {Number{2, -30}, 2, Number{1414213562373095, -30}},
+            {Number{-27}, 3, Number{-3}},
+            {Number{1}, 5, Number{1}},
+            {Number{-1}, 0, Number{1}},
+            {Number{5, -1}, 0, Number{0}},
```

Comment:
I think these calls are asking for the zeroth root of `-1` and `0.5` respectively?  If that reading is correct, a couple of non-authoritative) web sites I checked suggest that the zeroth root of any number (possibly excepting 1) is undefined:

- https://math.stackexchange.com/questions/962807/does-the-zeroth-root-exist
- https://www.quora.com/What-is-the-zeroth-root-of-any-number

So I'm wondering if the better answer for taking the zeroth root of a number is to throw?  I'm open to being wrong on this call...

Reply:
I went with Annex F of the C standard: https://cigix.me/c17#F.10.4.4.p1

> pow(−1, ±∞) returns 1.
> pow(x, +∞) returns +0 for |x| < 1.

Annex F is consistent with the IEEE floating point standards: https://cigix.me/c17#F.1.p1